### PR TITLE
[JUJU-3600] Remove 32-bit architecture support

### PR DESCRIPTION
**Pull request by SimonRichardson**
This is the first step in fixing the problematic arch supported list from LXD. Whilst investigating that bug, it seems that we still support 32-bit word sizes. Except that, we no longer want to support 32-bit word sizes with dqlite, as the number of jujud permutations explodes and so it's best to completely remove the support. We officially dropped support for 32-bit architectures over 3 years ago, although we didn't actively remove the code to drop support.

This brings in what the juju/utils arch support had, with the addition of a UnsupportedArches field, which will ensure that we can at least say what we use to support, but no longer do. This doesn't bring in the changes from [juju/utils PR](https://github.com/juju/utils/pull/330) as technically we don't need a new version of juju/utils sans arch package to do this.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/2013869

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]

File: core/arch/arches_test.go [hunk 1]
```diff
@@ -25,22 +24,45 @@ func (s archSuite) TestContains(c *gc.C) {
...
+		{"windows", "windows"},
+		{"amd64", "amd64"},
+		{"x86_64", "amd64"},
+		{"aarch64", "arm64"},
+		{"arm64", "arm64"},
+		{"ppc64el", "ppc64el"},
+		{"ppc64le", "ppc64el"},
+		{"s390x", "s390x"},
+		{"riscv64", "riscv64"},
+		{"risc", "riscv64"},
```
Comment (SimonRichardson) [inline core/arch/arches_test.go:54]: I'm unsure about this one, as we do move it to `riscv64`, I'm not a fan of dropping the word size, I think we should be a lot stricter.

File: core/arch/arches_test.go [hunk 2]
```diff
@@ -25,22 +24,45 @@ func (s archSuite) TestContains(c *gc.C) {
...
+	a := arch.HostArch()
+	c.Assert(arch.IsSupportedArch(a), jc.IsTrue)
+}
+
+func (s *archSuite) TestNormaliseArch(c *gc.C) {
+	for _, test := range []struct {
+		raw  string
+		arch string
+	}{
+		{"windows", "windows"},
```
Comment (SimonRichardson) [inline core/arch/arches_test.go:45]: I've just noticed, how is "windows" an arch?

File: core/arch/arches.go [hunk 3]
```diff
@@ -49,14 +48,70 @@ func (a Arches) String() string {
...
+	ARM64,
+	PPC64EL,
+	S390X,
+	RISCV64,
+}
+
+// UnsupportedArches records the machine architectures not supported by Juju.
+// Note: don't make const to prevent referencing it.
+var UnsupportedArches = []string{
+	"i386", "armhf",
```
Comment (hmlanigan) [inline core/arch/arches.go:72]: maybe include a few more in the list, the regexes for these include more variants.

File: core/arch/arches.go [hunk 4]
```diff
@@ -49,14 +48,70 @@ func (a Arches) String() string {
...
+
+// archREs maps regular expressions for matching
+// `uname -m` to architectures recognised by Juju.
+var archREs = []struct {
+	*regexp.Regexp
+	arch string
+}{
+	{Regexp: regexp.MustCompile("amd64|x86_64"), arch: AMD64},
+	{Regexp: regexp.MustCompile("aarch64"), arch: ARM64},
+	{Regexp: regexp.MustCompile("ppc64el|ppc64le"), arch: PPC64EL},
```
Comment (hmlanigan) [inline core/arch/arches.go:83]: Maybe add `ppc64` to the list.

From commit message: "Additionally, I've dropped support for ppc and we explicitly callout ppc64(le|el)." 
ppc64 was the listing in the arch package, which is still 64 bit, why drop it?

File: apiserver/facades/client/application/deployrepository.go [hunk 5]
```diff
@@ -531,7 +531,7 @@ func (v *deployFromRepositoryValidator) resolveCharm(curl *charm.URL, requestedO
 	// default arch, it means use any arch which isn't quite the same. So if we
 	// do get "all" we should see if there is a clean way to resolve it.
 	if resolvedOrigin.Platform.Architecture == "all" {
-		resolvedOrigin.Platform.Architecture = arch.ConstraintArch(modelCons, nil)
+		resolvedOrigin.Platform.Architecture = constraints.ConstraintArch(modelCons, nil)
```
Comment (hmlanigan) [inline apiserver/facades/client/application/deployrepository.go:534]: Suggest renaming to `constraints.Arch` instead. Reads easier.

File: caas/kubernetes/provider/k8s.go [hunk 6]
```diff
@@ -16,9 +16,9 @@ import (
 	jujuclock "github.com/juju/clock"
 	"github.com/juju/collections/set"
 	"github.com/juju/errors"
+	"github.com/juju/juju/core/arch"
```
Comment (hmlanigan) [inline caas/kubernetes/provider/k8s.go:19]: Wrong import stanza. Didn't the linter or SA check these? This happened in many of the files.

File: core/arch/arches_test.go (hunk 2 above)
Comment (hmlanigan) [reply core/arch/arches_test.go:45]: Remove it - we no longer support windows charms anyway

File: provider/dummy/environs.go [hunk 7]
```diff
@@ -1179,7 +1179,7 @@ func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constra
 	validator := constraints.NewValidator()
 	validator.RegisterUnsupported(unsupportedConstraints)
 	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
-	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL, arch.S390X})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.PPC64EL, arch.S390X})
```
Comment (hmlanigan) [inline provider/dummy/environs.go:1182]: Add risk5 while we're here?

**Review by hmlanigan — APPROVED** [review]
Should follow up and remove arch from utils, make it utils v4 so we don't accidentally start using it again. Especially with merges from 3.1 etc.

What are the odds there's a juju config running focal on i386 somewhere that might try to upgrade to juju 3.2? What happens if so?

**Comment by SimonRichardson** [comment]
@hmlanigan RE:

> Should follow up and remove arch from utils, make it utils v4 so we don't accidentally start using it again

I can't land it because of there is a ssh test race which I'm unable to fix atm https://github.com/juju/utils/pull/330

**Pull request merged**